interfaces/hotplug/udevadm.go

```diff
@@ -0,0 +1,126 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"bufio"
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+var udevadmBin = `udevadm`
+
+// RunUdevadm enumerates all devices by parsing 'udevadm info -e' command output and reports them
+// via devices channel. The devices channel gets closed to indicate that all devices were processed.
+// Parsing errors are reported to parseErrors channel. Fatal errors encountered when starting udevadm
+// are reported by the error return value.
+func RunUdevadm(devices chan<- *HotplugDeviceInfo, parseErrors chan<- error) error {
+	cmd := exec.Command(udevadmBin, "info", "-e")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		close(devices)
```

Comment:
Good idea. I evisioned it would be the responsibility of the caller (and I only considered closing devices channel important to signify end of processing), but it's cleaner that way, thanks for suggestion.